File: sennaar-rs/src/registry/platform.rs
#![allow(non_camel_case_types)]

use std::fmt::{Display, Formatter, Result as FmtResult};
use std::str::FromStr;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};


include!("../macross.rs");

ss_enum_wcustom! {
    Arch,
    i386, x86_64, aarch64, riscv64
}

ss_enum! {
    Endian, little, big
}

ss_enum_wcustom! {
    OS,
    windows, linux, macos, freebsd
}

ss_enum_wcustom! {
    LibC,
    msft, musl, glibc
}

#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[derive(JsonSchema)]
#[serde(tag = "$kind")]
pub enum PlatformSpecifierState<T> {
    Exact { value: T },
    Other,
    Any
}

impl<T: Display> PlatformSpecifierState<T> {
    fn write_with_other_and_any(
        &self,
        f: &mut Formatter<'_>,
        other: &str,
        any: &str,
    ) -> FmtResult {
        match self {
            PlatformSpecifierState::Exact { value } => write!(f, "{value}"),
            PlatformSpecifierState::Other => write!(f, "{other}"),
            PlatformSpecifierState::Any => write!(f, "{any}"),
        }
    }
}

impl<T: FromStr> PlatformSpecifierState<T> {
    pub fn parse_with_other_and_any(
        s: &str,
        other: &str,
        any: &str,
    ) -> Result<Self, T::Err> {
        if s == other {
            Ok(PlatformSpecifierState::Other)
        } else if s == any {
            Ok(PlatformSpecifierState::Any)
        } else {
            Ok(PlatformSpecifierState::Exact { value: s.parse()? })
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
#[derive(JsonSchema)]
pub struct Platform {
    pub arch: PlatformSpecifierState<Arch>,
    pub endian: Option<Endian>,
    pub os: PlatformSpecifierState<OS>,
    pub libc: PlatformSpecifierState<LibC>,
    pub custom: PlatformSpecifierState<String>,
}

pub const OTHER_ARCH: &'static str = "other_arch";
pub const OTHER_OS: &'static str = "other_os";
pub const OTHER_LIBC: &'static str = "other_libc";
pub const OTHER_CUSTOM: &'static str = "[other]";

pub const ANY_ARCH: &'static str = "any_arch";
pub const ANY_ENDIAN: &'static str = "any_endian";
pub const ANY_OS: &'static str = "any_os";
pub const ANY_LIBC: &'static str = "any_libc";
pub const ANY_CUSTOM: &'static str = "[any]";

impl Display for Platform {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        self.arch.write_with_other_and_any(f, OTHER_ARCH, ANY_ARCH)?;
        write!(f, "-")?;

        if let Some(endian) = &self.endian {
            write!(f, "{endian}-")?;
        } else {
            write!(f, "{ANY_ENDIAN}-")?;
        }

        self.os.write_with_other_and_any(f, OTHER_OS, ANY_OS)?;
        write!(f, "-")?;

        self.libc.write_with_other_and_any(f, OTHER_LIBC, ANY_LIBC)?;
        write!(f, "-")?;

        self.custom.write_with_other_and_any(f, OTHER_CUSTOM, ANY_CUSTOM)?;

        Ok(())
    }
}

impl FromStr for Platform {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.to_lowercase();
        let parts: Vec<&str> = s.splitn(5, '-').collect();
        if parts.len() != 5 {
            return Err("Platform string must have 5 parts".to_string());
        }

        let arch = PlatformSpecifierState::parse_with_other_and_any(parts[0], OTHER_ARCH, ANY_ARCH).unwrap();
        let endian = if parts[1] == ANY_ENDIAN {
            None
        } else {
            Some(parts[1].parse()?)
        };
        let os = PlatformSpecifierState::parse_with_other_and_any(parts[2], OTHER_OS, ANY_OS).unwrap();
        let libc = PlatformSpecifierState::parse_with_other_and_any(parts[3], OTHER_LIBC, ANY_LIBC).unwrap();
        let custom: PlatformSpecifierState<String> = PlatformSpecifierState::parse_with_other_and_any(parts[4], OTHER_CUSTOM, ANY_CUSTOM).unwrap();

        Ok(Platform {
            arch,
            endian,
            os,
            libc,
            custom,
        })
    }
}

File: sennaar-kt/src/main/kotlin/registry/platform.kt
package club.doki7.sennaar.registry

import club.doki7.sennaar.registry.Arch.AArch64
import club.doki7.sennaar.registry.Arch.Custom
import club.doki7.sennaar.registry.Arch.RiscV64
import club.doki7.sennaar.registry.Arch.X86
import club.doki7.sennaar.registry.Arch.X86_64
import club.doki7.sennaar.registry.Endian.Big
import club.doki7.sennaar.registry.Endian.Little
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

@Suppress("ClassName")
@Serializable(with = ArchSerializer::class)
sealed interface Arch {
    object X86 : Arch { override fun toString(): String = "x86" }
    object X86_64 : Arch { override fun toString(): String = "x86_64" }
    object AArch64 : Arch { override fun toString(): String = "aarch64" }
    object RiscV64 : Arch { override fun toString(): String = "riscv64" }
    data class Custom(val name: String) : Arch {
        override fun toString(): String = name
    }
}

fun parseArch(name: String): Arch {
    return when (name) {
        "x86" -> X86
        "x86_64" -> X86_64
        "aarch64" -> AArch64
        "riscv64" -> RiscV64
        else -> Custom(name)
    }
}

object ArchSerializer : KSerializer<Arch> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Arch", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): Arch {
        val archName = decoder.decodeString()
        return when (archName) {
            "x86" -> X86
            "x86_64" -> X86_64
            "aarch64" -> AArch64
            "riscv64" -> RiscV64
            else -> Custom(archName)
        }
    }

    override fun serialize(encoder: Encoder, value: Arch) {
        encoder.encodeString(value.toString())
    }
}

@Serializable
enum class Endian {
    @SerialName("little") Little,
    @SerialName("big") Big;

    override fun toString(): String {
        return when (this) {
            Little -> "little"
            Big -> "big"
        }
    }
}

fun parseEndian(name: String): Endian {
    return when (name) {
        "little" -> Little
        "big" -> Big
        else -> error("Unknown endian: $name")
    }
}

@Serializable(with = OSSerializer::class)
sealed interface OS {
    object Windows : OS { override fun toString(): String = "windows" }
    object Linux : OS { override fun toString(): String = "linux" }
    object MacOS : OS { override fun toString(): String = "macos" }
    class FreeBSD : OS { override fun toString(): String = "freebsd" }
    data class Custom(val name: String) : OS {
        override fun toString(): String = name
    }
}

fun parseOS(name: String): OS {
    return when (name) {
        "windows" -> OS.Windows
        "linux" -> OS.Linux
        "macos" -> OS.MacOS
        "freebsd" -> OS.FreeBSD()
        else -> OS.Custom(name)
    }
}

object OSSerializer : KSerializer<OS> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("OS", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): OS {
        val osName = decoder.decodeString()
        return when (osName) {
            "windows" -> OS.Windows
            "linux" -> OS.Linux
            "macos" -> OS.MacOS
            "freebsd" -> OS.FreeBSD()
            else -> OS.Custom(osName)
        }
    }

    override fun serialize(encoder: Encoder, value: OS) {
        encoder.encodeString(value.toString())
    }
}

@Serializable(with = LibcSerializer::class)
sealed interface LibC {
    object MSFT : LibC { override fun toString(): String = "msft" }
    object MUSL : LibC { override fun toString(): String = "musl" }
    object GLibC : LibC { override fun toString(): String = "glibc" }
    data class Custom(val name: String) : LibC {
        override fun toString(): String = name
    }
}

fun parseLibC(name: String): LibC {
    return when (name) {
        "msft" -> LibC.MSFT
        "musl" -> LibC.MUSL
        "glibc" -> LibC.GLibC
        else -> LibC.Custom(name)
    }
}

object LibcSerializer : KSerializer<LibC> {
    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("LibC", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): LibC {
        val libcName = decoder.decodeString()
        return when (libcName) {
            "msft" -> LibC.MSFT
            "musl" -> LibC.MUSL
            "glibc" -> LibC.GLibC
            else -> LibC.Custom(libcName)
        }
    }

    override fun serialize(encoder: Encoder, value: LibC) {
        encoder.encodeString(value.toString())
    }
}

@Serializable
sealed interface PlatformSpecifierState<T> {
    @SerialName("Exact")
    data class Exact<T>(val value: T) : PlatformSpecifierState<T>
    @SerialName("Other")
    class Other<T> : PlatformSpecifierState<T>
    @SerialName("Any")
    class Any<T> : PlatformSpecifierState<T>

    fun toStringWithOtherAndAny(other: String, any: String): String {
        return when (this) {
            is Exact -> value.toString()
            is Other -> other
            is Any -> any
        }
    }
}

fun<T> parsePlatformSpecifierState(
    value: String,
    valueParser: (String) -> T,
    other: String,
    any: String
): PlatformSpecifierState<T> {
    return when (value) {
        other -> PlatformSpecifierState.Other()
        any -> PlatformSpecifierState.Any()
        else -> PlatformSpecifierState.Exact(valueParser(value))
    }
}

const val OTHER_ARCH: String = "other_arch"
const val OTHER_OS: String = "other_os"
const val OTHER_LIBC: String = "other_libc"
const val OTHER_CUSTOM: String = "[other]"

const val ANY_ARCH: String = "any_arch"
const val ANY_ENDIAN: String = "any_endian"
const val ANY_OS: String = "any_os"
const val ANY_LIBC: String = "any_libc"
const val ANY_CUSTOM: String = "[any]"

@Serializable
data class Platform(
    var arch: PlatformSpecifierState<Arch>,
    var endian: Endian?,
    var os: PlatformSpecifierState<OS>,
    var libc: PlatformSpecifierState<LibC>,
    var custom: PlatformSpecifierState<String>
) {
    override fun toString(): String {
        return "${arch.toStringWithOtherAndAny(OTHER_ARCH, ANY_ARCH)}-" +
                "${endian?.toString() ?: ANY_ENDIAN}-" +
                "${os.toStringWithOtherAndAny(OTHER_OS, ANY_OS)}-" +
                "${libc.toStringWithOtherAndAny(OTHER_LIBC, ANY_LIBC)}-" +
                custom.toStringWithOtherAndAny(OTHER_CUSTOM, ANY_CUSTOM)
    }
}

fun parsePlatform(s: String): Platform {
    val s = s.lowercase()
    val parts = s.split(":", limit = 5)
    if (parts.size != 5) {
        error("Platform string must have 5 parts")
    }

    val arch: PlatformSpecifierState<Arch> = parsePlatformSpecifierState(
        parts[0],
        ::parseArch,
        OTHER_ARCH,
        ANY_ARCH
    )
    val endian = if (parts[1] == ANY_ENDIAN) {
        null
    } else {
        parseEndian(parts[1])
    }
    val os = parsePlatformSpecifierState(
        parts[2],
        ::parseOS,
        OTHER_OS,
        ANY_OS
    )
    val libc = parsePlatformSpecifierState(
        parts[3],
        ::parseLibC,
        OTHER_OS,
        ANY_OS
    )
    val custom = parsePlatformSpecifierState(
        parts[4],
        { it },
        OTHER_CUSTOM,
        ANY_CUSTOM
    )

    return Platform(arch, endian, os, libc, custom)
}
